### DISCUSSION TITLE
Memory leak when using `flatMapPrefix`/`prefixAndTail` with `Concat`

### DISCUSSION THREAD

**Discussion by queimadus** [apache/pekko#1566] [Q&A] [answered]
When using `flatMapPrefix` or `prefixAndTail` (or any other flow that closes over data) followed by `concat` the accumulated portion doesn't seem to be GC'd.

I reduced a problem I found into this snippet:

```scala
val s = Source
  .repeat(())
  .map(_ => ByteString('a' * 40000))
  .take(1000000)
  .flatMapPrefix(50000) { prefix => Flow[ByteString] } 

Source.empty
  .concatAllLazy(List.tabulate(30000)(_ => s): _*)
  .runWith(Sink.ignore).onComplete(println(_))  
```

If we don't use concat, it runs fine:

```scala
  Source
    .repeat(s)
    .take(30000)
    .flatMapConcat(x => x)
    .runWith(Sink.ignore)
    .onComplete(println(_))
```

**Comment by He-Pin** [comment]
@queimadus What about `wrapping` it into a `Source. lazySource` ? the concatAll will materialize all sources

**Reply by He-Pin** [reply]
I added a pr to null out the unused ref, but that will just help a little.
https://github.com/apache/pekko/pull/1615

**Reply by queimadus** [reply]
I tried the lazy variations as well and got the same behaviour. In the example above it uses `concatAllLazy` which shouldn't materialize all sources eagerly.

The problem doesn't seem to be as much related to eager evaluation but instead not being able to fully clean-up the source resources once it's used up.

In my real use-case I've changed some things around to be able to implement it with a `flatMapConcat` instead of using `concat` variants, and it solved the problem.

At the time I grabbed these screenshots (from my real-use case and not this MRE, but the profile should be similar):
![image](https://github.com/user-attachments/assets/f891bcbd-d45a-4faf-b613-3a87c0dcf0f4)
![image](https://github.com/user-attachments/assets/23b8edc7-4e85-4e1c-9077-a799f44e45f4)

**Reply by He-Pin** [reply]
thanks for sharing the pictures,  did you looked into the code? Seems there are some strong references. 

I was want to set the consumed inlet to null   ,but that doesn't work, because it's a vector.

**Reply by He-Pin** [reply]
```
Source.empty
  .concatAllLazy(List.tabulate(30000)(_ => Source.lazySource(_ => s): _*)
  .runWith(Sink.ignore).onComplete(println(_))  
```

Will help I think.

**Reply by queimadus** [reply]
Think I've tried that too but I'll confirm. 

I'll spin up a new project, create some real mem dump analysis from these examples, test the changes you've suggested, and I'll post them here later 👍

**Reply by He-Pin** [reply]
I think we should clean up the `portToConn` , set it `portToConn(i)` to null when it beeen consumed, wdyt

**Reply by queimadus** [reply]
Ran this code https://github.com/queimadus/pekko-leak-example with 100M of heap. Both tests using concat ran out of heap.

Using the same inner source in all examples:

```
  val s = Source
    .repeat(())
    .map(_ => ByteString('a' * 400000))
    .take(1000000)
    .flatMapPrefix(50000) { prefix => Flow[ByteString] }
 ```

-----------------------------------

```
Source.empty
  .concatAllLazy(List.tabulate(30000)(_ => s): _*)
  .runWith(Sink.ignore).onComplete(println(_))  
```

<img width="1840" alt="Screenshot 2024-12-21 at 15 48 37" src="https://github.com/user-attachments/assets/1c35e3f9-0cee-4ab7-ba2b-3cb29787a3ef" />
<img width="1262" alt="Screenshot 2024-12-21 at 15 53 05" src="https://github.com/user-attachments/assets/10d93402-6a6b-475a-88f9-5fd8e6895bbd" />

-------------------------------

```
Source
    .repeat(s)
    .take(30000)
    .flatMapConcat(x => x)
    .runWith(Sink.ignore)
    .onComplete(println(_))
```

<img width="1296" alt="Screenshot 2024-12-21 at 15 52 27" src="https://github.com/user-attachments/assets/5f64fd09-fec9-48da-abaa-3f1612677243" />
<img width="1224" alt="Screenshot 2024-12-21 at 15 53 26" src="https://github.com/user-attachments/assets/b02fd74c-66fb-4188-8ba1-2a4051e3ccd7" />

-------------------------------

```
  Source.empty
    .concatAllLazy(List.tabulate(30000)(_ => Source.lazySource(() => s)): _*)
    .runWith(Sink.ignore).onComplete(println(_))
```

<img width="1299" alt="Screenshot 2024-12-21 at 16 07 31" src="https://github.com/user-attachments/assets/092ceee8-963a-43ea-8cdd-636ffa5ca092" />
<img width="994" alt="Screenshot 2024-12-21 at 16 07 12" src="https://github.com/user-attachments/assets/666ea1ae-3b90-4ff3-ad15-5e0c8a43da62" />

**Reply by queimadus** [reply]
> I think we should clean up the `portToConn` , set it `portToConn(i)` to null when it beeen consumed, wdyt

I'm not familiar with the implementation details at that level, but if there's a specific change you want to try out, I can test it out myself.

**Reply by He-Pin** [reply]
flatMapPrefix(50000) will have 50000 vector which contains the bytestring.

**Reply by He-Pin** [reply]
![Screenshot_20241222_002531_Calculator.jpg](https://github.com/user-attachments/assets/b7f3413c-3999-42aa-ba52-f9a48e36aa0e)

Will it still happen when you change from 50000 to 1?

**Reply by queimadus** [reply]
Correct, it was intended. Just a way to generate a big chunk of memory that should be GC'd.  

>Will it still happen when you change from 50000 to 1?

I'd say so, just slower.

In my real use case, I needed to have this 4Mb chunks in mem, for multiple streams that would be fed into a concat (but should be fine since they could be GC'd)

**Reply by He-Pin** [reply]
I tooked a while of the code in the graph stage , where there is an array of input handers and connections, I think we should clean that array slot after a stage is completed, then we need more changes eg when the slot is null, we should treat the connection is closed where it's now protected by  flags.

I did not tried that today, would you take a try? I mean clone the project and made the changes to see if it works.

**Reply by queimadus** [reply]
I can take a look at it somewhere in the next few days!

**Reply by He-Pin** [reply]
Me too, I will take your repo as a starting point tomorrow.

**Reply by queimadus** [reply]
Tested it with a very naive change, just to check if it would make any difference in the test example, and got no difference in heap behavior. Maybe too much of a naive change 😅  

```diff
diff --git a/stream/src/main/scala/org/apache/pekko/stream/scaladsl/Graph.scala b/stream/src/main/scala/org/apache/pekko/stream/scaladsl/Graph.scala
index bf28108..aef5833 100755
--- a/stream/src/main/scala/org/apache/pekko/stream/scaladsl/Graph.scala
+++ b/stream/src/main/scala/org/apache/pekko/stream/scaladsl/Graph.scala
@@ -1349,6 +1349,7 @@ final class Concat[T](val inputPorts: Int) extends GraphStage[UniformFanInShape[
             }
 
             override def onUpstreamFinish() = {
+              connDelete(i)
               if (idx == activeStream) {
                 activeStream += 1
                 // Skip closed inputs
diff --git a/stream/src/main/scala/org/apache/pekko/stream/stage/GraphStage.scala b/stream/src/main/scala/org/apache/pekko/stream/stage/GraphStage.scala
index b186308..e09f123 100644
--- a/stream/src/main/scala/org/apache/pekko/stream/stage/GraphStage.scala
+++ b/stream/src/main/scala/org/apache/pekko/stream/stage/GraphStage.scala
@@ -483,6 +483,10 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
 
   private def conn(in: Inlet[_]): Connection = portToConn(in.id)
   private def conn(out: Outlet[_]): Connection = portToConn(out.id + inCount)
+  def connDelete(in: Inlet[_]): Unit = {
+    portToConn(in.id) = null
+    println((portToConn.count(_ eq null), portToConn.size))
+  }
 
   /**
    * Retrieves the current callback for the events on the given [[Outlet]]

```

**Reply by He-Pin** [reply]
Thanks, Another point of leak is the AsyncCallback, where is actually a queue behind of it, i mean the mailbox of the interpreter's backing actor.

**Reply by He-Pin** [reply, marked as answer]
@queimadus https://github.com/apache/pekko/issues/1652

**Comment by He-Pin** [comment]
@queimadus thanks , it's now fixed
